### PR TITLE
Inter-vPBX calls through Friends

### PR DESCRIPTION
**Pull request by cruzccl**
This PR aims to add inter-vPBX calls using already existing Friends.

**Note**: only calls to extensions of destination vPBX are allowed.

### CODE REVIEW

File: asterisk/agi/src/Agi/Action/FriendCallAction.php [hunk 1]
```diff
@@ -66,6 +66,13 @@ public function process()
         // Some verbose dolan pls
         $this->agi->notice("Preparing call to %s through friend <cyan>%s</cyan>", $number, $friend);
 
+        // Intervpbx friends can only call to valid extensions in destination company
+        if ($friend->getDirectConnectivity() == 'intervpbx' &&
```
Comment (Kaian) [inline asterisk/agi/src/Agi/Action/FriendCallAction.php]: Use FriendInterface::DIRECTCONNECTIVITY_INTERVPBX constant here

Also strict comparison (===) would be recommended

File: library/Ivoz/Ast/Domain/Service/PsEndpoint/UpdateByFriend.php [hunk 2]
```diff
@@ -65,6 +65,11 @@ public function execute(FriendInterface $entity)
             ? $entity->getFromDomain()
             : $entity->getDomain()->getDomain();
 
+        // Disable directMedia for intervpbx friends
+        if ($entity->getDirectConnectivity() == 'intervpbx') {
```
Comment (Kaian) [inline library/Ivoz/Ast/Domain/Service/PsEndpoint/UpdateByFriend.php]: Use interface constant here (and strict comparison)

File: scheme/app/DoctrineMigrations/Version20190408170724.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+{
+    /**
+     * @param Schema $schema
+     */
+    public function up(Schema $schema)
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() !== 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+
+        $this->addSql('ALTER TABLE Friends ADD interCompanyId INT UNSIGNED DEFAULT NULL, CHANGE directConnectivity directConnectivity VARCHAR(255) DEFAULT \'yes\' NOT NULL COMMENT \'[enum:yes|no|intervpbx]\'');
```
Comment (Kaian) [inline scheme/app/DoctrineMigrations/Version20190408170724.php]: 255 chars seems a bit excessive for a field with less than 10 possible characters.

**Comment by ironArt3mis** [comment]
Rebase completed.

**Comment by Kaian** [comment]
Should we squash some of this PR commits?

File: library/Ivoz/Provider/Domain/Model/Friend/Friend.php [hunk 4]
```diff
@@ -51,27 +52,34 @@ public function __toString()
...
+            // Force DDI In mode
+            $this->setDdiIn(FriendInterface::DDIIN_YES);
+            // Set From Domain from target company
+            $this->setFromDomain($this->getInterCompany()->getDomainUsers());
+        } else {
+            $this->setInterCompany(null);
+        }
+
+        // Validate Name format
         Assertion::regex(
```
Comment (mmadariaga) [inline library/Ivoz/Provider/Domain/Model/Friend/Friend.php]: Any reason not to validate the name into it's setter?

File: library/Ivoz/Provider/Domain/Model/Friend/Friend.php [hunk 5]
```diff
@@ -220,4 +228,34 @@ public function getOutgoingDdi()
 
         return $ddi;
     }
+
+
+    /**
+     * @return string
+     * @throws \Exception
+     */
+    public function getInterCompanyName()
```
Comment (mmadariaga) [inline library/Ivoz/Provider/Domain/Model/Friend/Friend.php]: This method should not be public, plus you can enforce return type (:string)

File: library/Ivoz/Provider/Domain/Model/Friend/Friend.php [hunk 6]
```diff
@@ -3,6 +3,7 @@
 namespace Ivoz\Provider\Domain\Model\Friend;
 
 use Assert\Assertion;
+use Assert\AssertionFailedException;
```
Comment (mmadariaga) [inline library/Ivoz/Provider/Domain/Model/Friend/Friend.php]: Unused

File: asterisk/agi/src/Agi/Action/FriendCallAction.php [hunk 7]
```diff
@@ -66,6 +66,13 @@ public function process()
         // Some verbose dolan pls
         $this->agi->notice("Preparing call to %s through friend <cyan>%s</cyan>", $number, $friend);
 
+        // Intervpbx friends can only call to valid extensions in destination company
+        if ($friend->getDirectConnectivity() === FriendInterface::DIRECTCONNECTIVITY_INTERVPBX &&
```
Comment (mmadariaga) [inline asterisk/agi/src/Agi/Action/FriendCallAction.php]: Could we name conditions here? Set them into a variable or  add a couple of methods to Friend as well

**Pull request merged**